File: packages/lynx-audio/ios/AudioModule.swift
import Foundation
import AVFoundation
import Lynx

/// Audio recording + playback bridge.
///
/// JS usage (via the `@sigx/lynx-audio` shim, not directly):
///
///   NativeModules.Audio.play(uri, opts, cb)
///   NativeModules.Audio.pausePlayer(id, cb)
///   NativeModules.Audio.startRecording(opts, cb)
///   ...
///
/// State lives in `AudioPlayerRegistry` / `AudioRecorderRegistry`. The
/// module class is a thin dispatcher — Lynx allocates a fresh instance
/// per call so it intentionally holds no per-handle state itself.
/// Async events (`onEnd`, `onMeter`) flow through `AudioEventBus`.
@objc class AudioModule: NSObject, LynxModule {

    @objc static var name: String { "Audio" }

    @objc static var methodLookup: [String: String] {
        [
            "play":               NSStringFromSelector(#selector(play(_:options:callback:))),
            "preload":            NSStringFromSelector(#selector(preload(_:callback:))),
            "pausePlayer":        NSStringFromSelector(#selector(pausePlayer(_:callback:))),
            "resumePlayer":       NSStringFromSelector(#selector(resumePlayer(_:callback:))),
            "stopPlayer":         NSStringFromSelector(#selector(stopPlayer(_:callback:))),
            "seekPlayer":         NSStringFromSelector(#selector(seekPlayer(_:seconds:callback:))),
            "setPlayerVolume":    NSStringFromSelector(#selector(setPlayerVolume(_:volume:callback:))),
            "getPlayerStatus":    NSStringFromSelector(#selector(getPlayerStatus(_:callback:))),
            "startRecording":     NSStringFromSelector(#selector(startRecording(_:callback:))),
            "pauseRecording":     NSStringFromSelector(#selector(pauseRecording(_:callback:))),
            "resumeRecording":    NSStringFromSelector(#selector(resumeRecording(_:callback:))),
            "stopRecording":      NSStringFromSelector(#selector(stopRecording(_:callback:))),
            "setMeterSubscribed": NSStringFromSelector(#selector(setMeterSubscribed(_:subscribed:callback:))),
            "requestPermission":  NSStringFromSelector(#selector(requestPermission(_:))),
            "getPermissionStatus": NSStringFromSelector(#selector(getPermissionStatus(_:))),
        ]
    }

    required override init() { super.init() }
    required init(param: Any) { super.init() }

    // MARK: - Playback

    @objc func play(_ source: String?, options: [String: Any]?, callback: LynxCallbackBlock?) {
        guard let source = source, !source.isEmpty else {
            callback?(["error": "Missing source"])
            return
        }
        let volume = (options?["volume"] as? NSNumber)?.doubleValue ?? 1.0
        let loop = (options?["loop"] as? NSNumber)?.boolValue ?? false
        let rate = (options?["rate"] as? NSNumber)?.doubleValue ?? 1.0
        let result = AudioPlayerRegistry.shared.create(
            source: source, volume: volume, loop: loop, rate: rate
        )
        callback?(result)
    }

    @objc func preload(_ source: String?, callback: LynxCallbackBlock?) {
        guard let source = source, !source.isEmpty else {
            callback?(["error": "Missing source"])
            return
        }
        callback?(AudioPlayerRegistry.shared.preload(source: source))
    }

    @objc func pausePlayer(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.pause(id: id.int64Value))
    }

    @objc func resumePlayer(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.resume(id: id.int64Value))
    }

    @objc func stopPlayer(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.stop(id: id.int64Value))
    }

    @objc func seekPlayer(_ id: NSNumber, seconds: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.seek(id: id.int64Value, seconds: seconds.doubleValue))
    }

    @objc func setPlayerVolume(_ id: NSNumber, volume: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.setVolume(id: id.int64Value, volume: volume.doubleValue))
    }

    @objc func getPlayerStatus(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioPlayerRegistry.shared.status(id: id.int64Value))
    }

    // MARK: - Recording

    @objc func startRecording(_ options: [String: Any]?, callback: LynxCallbackBlock?) {
        let path = options?["outputPath"] as? String
        let format = (options?["format"] as? String) ?? "m4a"
        let sampleRate = (options?["sampleRate"] as? NSNumber)?.doubleValue ?? 44100
        let channels = (options?["channels"] as? NSNumber)?.intValue ?? 1

        // iOS demands the mic permission be granted before AVAudioRecorder
        // can `record()` — if not granted yet, surface that error rather
        // than silently producing a 0-byte file.
        let status = AVAudioSession.sharedInstance().recordPermission
        guard status == .granted else {
            callback?(["error": "Microphone permission not granted"])
            return
        }

        callback?(AudioRecorderRegistry.shared.start(
            outputPath: path, format: format, sampleRate: sampleRate, channels: channels
        ))
    }

    @objc func pauseRecording(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioRecorderRegistry.shared.pause(id: id.int64Value))
    }

    @objc func resumeRecording(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioRecorderRegistry.shared.resume(id: id.int64Value))
    }

    @objc func stopRecording(_ id: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioRecorderRegistry.shared.stop(id: id.int64Value))
    }

    @objc func setMeterSubscribed(_ id: NSNumber, subscribed: NSNumber, callback: LynxCallbackBlock?) {
        callback?(AudioRecorderRegistry.shared.setMeterSubscribed(
            id: id.int64Value, subscribed: subscribed.boolValue
        ))
    }

    // MARK: - Permission

    @objc func requestPermission(_ callback: LynxCallbackBlock?) {
        let session = AVAudioSession.sharedInstance()
        switch session.recordPermission {
        case .granted:
            callback?(["status": "granted", "canAskAgain": false])
        case .denied:
            // iOS won't re-prompt on second call — user must visit Settings.
            callback?(["status": "denied", "canAskAgain": false])
        case .undetermined:
            session.requestRecordPermission { granted in
                callback?([
                    "status": granted ? "granted" : "denied",
                    "canAskAgain": false,
                ])
            }
        @unknown default:
            callback?(["status": "denied", "canAskAgain": false])
        }
    }

    @objc func getPermissionStatus(_ callback: LynxCallbackBlock?) {
        let session = AVAudioSession.sharedInstance()
        switch session.recordPermission {
        case .granted:
            callback?(["status": "granted", "canAskAgain": false])
        case .denied:
            callback?(["status": "denied", "canAskAgain": false])
        case .undetermined:
            callback?(["status": "undetermined", "canAskAgain": true])
        @unknown default:
            callback?(["status": "denied", "canAskAgain": false])
        }
    }
}

File: packages/lynx-audio/android/com/sigx/audio/AudioModule.kt
package com.sigx.audio

import android.content.Context
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.ReadableMap
import com.sigx.permissions.PermissionHelper

/**
 * Audio recording + playback bridge.
 *
 * JS usage (via the `@sigx/lynx-audio` shim, not directly):
 *
 *   NativeModules.Audio.play(uri, opts, cb)
 *   NativeModules.Audio.pausePlayer(id, cb)
 *   NativeModules.Audio.startRecording(opts, cb)
 *   ...
 *
 * State lives in [AudioPlayerRegistry] / [AudioRecorderRegistry]. Async
 * events (`onEnd`, `onMeter`) flow through [AudioEventBus] → per-LynxView
 * [AudioPublisher] → JS `GlobalEventEmitter`. Microphone permission is
 * delegated to `@sigx/lynx-permissions`' `PermissionHelper` (`"microphone"`
 * → `RECORD_AUDIO`).
 */
class AudioModule(context: Context) : LynxModule(context) {

    // MARK: - Playback

    @LynxMethod
    fun play(source: String?, options: ReadableMap?, callback: Callback?) {
        if (source.isNullOrEmpty()) {
            callback?.invoke(errorMap("Missing source"))
            return
        }
        val volume = if (options?.hasKey("volume") == true) options.getDouble("volume") else 1.0
        val loop = if (options?.hasKey("loop") == true) options.getBoolean("loop") else false
        val rate = if (options?.hasKey("rate") == true) options.getDouble("rate") else 1.0
        callback?.invoke(AudioPlayerRegistry.create(mContext, source, volume, loop, rate))
    }

    @LynxMethod
    fun preload(source: String?, callback: Callback?) {
        if (source.isNullOrEmpty()) {
            callback?.invoke(errorMap("Missing source"))
            return
        }
        callback?.invoke(AudioPlayerRegistry.preload(mContext, source))
    }

    @LynxMethod
    fun pausePlayer(id: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.pause(id.toLong()))
    }

    @LynxMethod
    fun resumePlayer(id: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.resume(id.toLong()))
    }

    @LynxMethod
    fun stopPlayer(id: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.stop(id.toLong()))
    }

    @LynxMethod
    fun seekPlayer(id: Double, seconds: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.seek(id.toLong(), seconds))
    }

    @LynxMethod
    fun setPlayerVolume(id: Double, volume: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.setVolume(id.toLong(), volume))
    }

    @LynxMethod
    fun getPlayerStatus(id: Double, callback: Callback?) {
        callback?.invoke(AudioPlayerRegistry.status(id.toLong()))
    }

    // MARK: - Recording

    @LynxMethod
    fun startRecording(options: ReadableMap?, callback: Callback?) {
        val path = if (options?.hasKey("outputPath") == true) options.getString("outputPath") else null
        val format = if (options?.hasKey("format") == true) options.getString("format") ?: "m4a" else "m4a"
        val sampleRate = if (options?.hasKey("sampleRate") == true) options.getDouble("sampleRate").toInt() else 44100
        val channels = if (options?.hasKey("channels") == true) options.getDouble("channels").toInt() else 1

        // MediaRecorder requires RECORD_AUDIO to be granted before `start()`
        // — surface a clear error rather than letting the call throw deep
        // inside MediaRecorder's state machine.
        val status = PermissionHelper.checkPermission(mContext, "microphone")
        if (status.getString("status") != "granted") {
            callback?.invoke(errorMap("Microphone permission not granted"))
            return
        }

        callback?.invoke(AudioRecorderRegistry.start(mContext, path, format, sampleRate, channels))
    }

    @LynxMethod
    fun pauseRecording(id: Double, callback: Callback?) {
        callback?.invoke(AudioRecorderRegistry.pause(id.toLong()))
    }

    @LynxMethod
    fun resumeRecording(id: Double, callback: Callback?) {
        callback?.invoke(AudioRecorderRegistry.resume(id.toLong()))
    }

    @LynxMethod
    fun stopRecording(id: Double, callback: Callback?) {
        callback?.invoke(AudioRecorderRegistry.stop(id.toLong()))
    }

    @LynxMethod
    fun setMeterSubscribed(id: Double, subscribed: Boolean, callback: Callback?) {
        callback?.invoke(AudioRecorderRegistry.setMeterSubscribed(id.toLong(), subscribed))
    }

    // MARK: - Permission

    @LynxMethod
    fun requestPermission(callback: Callback?) {
        PermissionHelper.requestPermission(mContext, "microphone") { result ->
            callback?.invoke(result)
        }
    }

    @LynxMethod
    fun getPermissionStatus(callback: Callback?) {
        callback?.invoke(PermissionHelper.checkPermission(mContext, "microphone"))
    }

    private fun errorMap(message: String): com.lynx.react.bridge.JavaOnlyMap {
        val map = com.lynx.react.bridge.JavaOnlyMap()
        map.putString("error", message)
        return map
    }
}
